### ISSUE TITLE
Model database missing field separator in full_name field in 2.0-rc3

### ISSUE THREAD

**Issue by jceresini** [GoogleCloudPlatform/forseti-security#1703]
The model table has a field named `full_name`, which is apparently the `parent_type_name` and `type_name` fields concatenated. (Side note, the `type_name` field is the `type` and `name` fields concatenated). The `full_name` field is missing the field separator. Here is an example of some resources in the model database:

```
> select full_name, type_name, parent_type_name, name, type from `4c7bbdfe47888e85289671d594ed8253_resources` where type='network'\G
*************************** 1. row ***************************
       full_name: project/forseti-testxsubnetwork/6917533303373712976/
       type_name: network/6917533303373712976
parent_type_name: project/forseti-testxsub
            name: 6917533303373712976
            type: network
1 row in set (0.0393 sec)
```

```
select full_name, type_name, parent_type_name, name, type from `cb910b2f96d36ce76ef2a6447c96c4da_resources` where type='subnetwork' limit 1\G
*************************** 1. row ***************************
       full_name: project/forseti-testxsubnetwork/1482312637655436839/
       type_name: subnetwork/1482312637655436839
parent_type_name: project/forseti-testx
            name: 1482312637655436839
            type: subnetwork
1 row in set (0.0397 sec)
```

In this case I created 2 projects (forseti-testx, and forseti-testxsub) to demonstrate that this could, in extremely rare cases, create overlapping `full_name` identifiers. However, this looks like its just typo when generating the full_name. That field also has a trailing '/' so im guessing the field separator was just misplaced.

I'm also curious why the database has so much redundant data. Is there a technical reason to store the data like that?

**Comment by jceresini** [comment]
Oh, one other thing I think is worth mentioning. I think it could be valuable to store the name of the api a resource was retrieved from. For example, `compute` for instances, disks, etc. The concern is that google products may end up have overlapping type names across APIs. Possible example could be a `table` type in BigQuery and in BigTable.

**Comment by joecheuk** [comment]
Hi @jceresini , all full name ends with a '/' that's why in the code we have '{parent_full_name}{resource_type_name}/'. I ran the same query on my environment and it seems like I am able to get a proper full name. Is there any reason why the full_names you have don't contain an organization? Forseti intended to run on an Org level and inventory data within that Org.

`
mysql> select full_name, type_name, parent_type_name, name, type from c67f9bf44ca488a1ee452738cb0cca64_resources where type='network' limit 1\G;
*************************** 1. row ***************************
       full_name: organization/660570133860/project/policyscanner/network/1071927039411978409/
       type_name: network/1071927039411978409
parent_type_name: project/policyscanner-henry
            name: 1071927039411978409
            type: network
1 row in set (0.01 sec)
`

**Comment by jceresini** [comment]
I think I see whats happening here. You're running forseti on an org, while I'm running it on a project. So first the org get its `full_name` generated with:

`to_full_resource_name('', 'organization/1234567890')` resulting in `organization/1234567890/`
Then the project is discovered and its full name is generated with
`to_full_resource_name('organization/1234567890/', 'project/project_name')` resulting in `organization/1234567890/project/project_name/`
Then finally the network full name is generated using:
`to_full_resource_name('organization/1234567890/project/project_name/', 'network/1234567890')` resulting in `organization/1234567890/project/project_name/network/1234567890/`

I would think a better solution would be to special-case the no-parent case of `to_full_resource_name` to return the second argument as-is, otherwise concatenate both with a separator. Or use something like os.path.join() to join the path-like structure of a full_resource_name. Assuming that a parent resource ends in a slash, seems like a poor choice here.

**Comment by jceresini** [comment]
Given the way it currently works, this appears to be the actual problem (in google/cloud/forseti/services/model/importer/importer.py):

```
    def _convert_project(self, project):
        """Convert a project to a database object.

        Args:
            project (object): Project to store.
        """

        data = project.get_resource_data()
        if self._is_root(project):
            parent, type_name = None, self._type_name(project)
            full_res_name = type_name
        else:
            parent, full_res_name, type_name = self._full_resource_name(
                project)
```

If the project is the root, this does not return the project with a trailing `/`. By contrast, the org full_name is generated as follows:

```
    def _convert_organization(self, organization):
       [...]
       org = self.dao.TBL_RESOURCE(
            full_name=to_full_resource_name('', type_name),
```

So perhaps the solution is to update both the folder and project functions to call to_full_resource_name with `('',type_name)` if they are root. I'll put together a pull request for those changes

**Comment by joecheuk** [comment]
Hi @jceresini , we do plan to allow users to install without org access (project/folder level) but in order to do that, the user will have to provide a "stub" root in the forseti-security/configs/forseti_conf_server.yaml file. Can you try updating field "**root_resource_id**" to some stub value (e.g. organization/1234567890), then run the `forseti server configuration reload` command to reload the configuration and start the inventory process over again see if the data model built on top of the new inventory contains the correct full name?

**Comment by jceresini** [comment]
I have my `root_resource_id` set to `projects/[my_project]`. That's what resulted in the above output for 2 different projects I've run it with. So the submitted pull request fixes an issue I've run into, which is demonstrated above.

**Comment by joecheuk** [comment]
Hi @jceresini , the intention of having full name is to contain the hierarchy of the resource in the organization, therefore the stub root has to be prefix with **organizations/** and preferably '**organizations/YOUR_ORG_ID**' so the generated full name actually make sense when you look at it.

**Comment by jceresini** [comment]
I think it might be useful to have a discussion with the forseti team regarding how we would like to use forseti.

The server config file allows non-org root_resource_ids. The model importer has conditionals in the `_convert_project` and `_convert_folder` functions to see if they are the root resource, and in those conditionals they create a full_resource_name that does not have a trailing slash, which breaks assumptions made in other parts of forseti.

**Comment by joecheuk** [comment]
Thanks for fixing this issue, this will be included in the release after the next one (v2.0.2).
